File: src/ios/OcrPluginManager.swift
//
//  OcrPluginManager.swift
//  IpcOcrSample
//
//  Created by 平山 裕也 on 2024/05/13.
//

import UIKit
import DriverCardOCRLibrary

// MARK: - Protocol
protocol OnOcrPluginManagerDelegate {
    // [errorCode]
    // 0: 成功
    // -1: 失敗
    // -2: 権限エラー
    func onResult(sender: OcrPluginManager, code: Int, result: OcrResultInfo?)
}

// MARK: - InnerClass
class OcrResultInfo {
    
    // MARK: - Member
    var mCardType: Int = 0
    var mName: String = ""
    var mAddress: String = ""
    var mBirthdate: String = ""
    var mGender: String = ""
    
}

class OcrPluginManager: DriverCardOCRDelegate {
    
    // MARK: - Define
    static let CODE_SUCCESS = 0
    static let CODE_FAILED = -1
    static let CODE_AUTHROIZE = -2

    static let CARDTYPE_MENKYOSYO = 1
    static let CARDTYPE_MYNUMBER = 2
    static let CARDTYPE_ZAIRYU = 3
    
    // MARK: - Member
    private var mDelegate: OnOcrPluginManagerDelegate? = nil
    
    // MARK: - DriverCardOCRDelegate
    func didScaned(result: [(String, String)], scanType: DriverCardOCRLibrary.SCAN_TYPE) {
        // 解析結果、result:項目名,項目内容のリスト　scanType:カードタイプ 0:運転免許証 1:マイナンバーカード 2:在留カード
        var info = OcrResultInfo()
        switch scanType {
            case SCAN_TYPE.DriverCard:
            info.mCardType = OcrPluginManager.CARDTYPE_MENKYOSYO
            result.forEach { set in
                if (set.0.compare("氏名") == .orderedSame) {
                    info.mName = set.1
                }
                if (set.0.compare("住所") == .orderedSame) {
                    info.mAddress = set.1
                }
                if (set.0.compare("生日") == .orderedSame) {
                    info.mBirthdate = set.1.replacingOccurrences(of: "生", with: "")
                }
            }
                break
            case SCAN_TYPE.ZairyuCard:
                info.mCardType = OcrPluginManager.CARDTYPE_ZAIRYU
                result.forEach { set in
                    if (set.0.compare("氏名") == .orderedSame) {
                        info.mName = set.1
                    }
                    if (set.0.compare("住所") == .orderedSame) {
                        info.mAddress = set.1
                    }
                    if (set.0.compare("生年月日") == .orderedSame) {
                        info.mBirthdate = set.1.replacingOccurrences(of: "生", with: "")
                    }
                    if (set.0.compare("性別") == .orderedSame) {
                        info.mGender = set.1
                    }
                }
                break
            case SCAN_TYPE.MyNumberCard:
                info.mCardType = OcrPluginManager.CARDTYPE_MYNUMBER
                var address1 = ""
                var address2 = ""
                result.forEach { set in
                    if (set.0.compare("氏名") == .orderedSame) {
                        info.mName = set.1
                    }
                    if (set.0.compare("住所①") == .orderedSame) {
                        address1 = set.1
                    }
                    if (set.0.compare("住所②") == .orderedSame) {
                        address2 = set.1
                    }
                    if (set.0.compare("生日") == .orderedSame) {
                        info.mBirthdate = set.1.replacingOccurrences(of: "生", with: "")
                    }
                    if (set.0.compare("性別") == .orderedSame) {
                        info.mGender = set.1
                    }
                }
                info.mAddress = address1 + address2
                break
            default:
                mDelegate?.onResult(sender: self, code: OcrPluginManager.CODE_FAILED, result: nil)
                return
        }
        mDelegate?.onResult(sender: self, code: OcrPluginManager.CODE_SUCCESS, result: info)
    }
    
    func canceled() {
        mDelegate?.onResult(sender: self, code: OcrPluginManager.CODE_FAILED, result: nil)
    }
    
    func staredProcessing() {
    }
    
    func endProcessing() {
    }
    
    // MARK: - Accesser
    func start(parent: UIViewController, delegate: OnOcrPluginManagerDelegate?) {
        UIApplication.shared.statusBarOrientation = .portrait

        mDelegate = delegate
        DriverCardOCR.shared.delegate = self
        DriverCardOCR.shared.parentController = parent
        DriverCardOCR.shared.doScanAutoCard()
    }
    
    func stop() {
        mDelegate = nil

        UIApplication.shared.statusBarOrientation = .landscapeLeft
    }
    
    
}

File: src/android/plugin/OcrPluginManager.java
package jp.bluememe.plugin.ocr.plugin;

import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;

import java.io.Serializable;

import jp.co.ip_consulting.drivercardocrlibrary.DriverCardOCR;
import kotlin.Unit;
import kotlin.jvm.functions.Function3;

public class OcrPluginManager {

    // MARK: - Define
    public static final int CODE_SUCCESS = 0;
    public static final int CODE_FAILED = -1;
    public static final int CODE_AUTHROIZE = -2;

    public static final int CARDTYPE_MENKYOSYO = 1;
    public static final int CARDTYPE_MYNUMBER = 2;
    public static final int CARDTYPE_ZAIRYU = 3;

    // MARK: - Member
    private OnOcrPluginManagerListener mListener = null;

    // MARK: - Listener
    public interface OnOcrPluginManagerListener {
        void onResult(OcrPluginManager sender, int code, OcrPluginManager.OcrResultInfo result);
    }

    // MARK: - Accesser
    public void start (AppCompatActivity activity, OnOcrPluginManagerListener listener) {
        mListener = listener;
        Permission.checkPermission(activity, new OnPermissionListener() {
            @Override
            public void onResult(boolean result) {
                if (result) {
                    doStartScan(activity);
                } else {
                    if (mListener != null) {
                        mListener.onResult(OcrPluginManager.this, CODE_AUTHROIZE, null);
                    }
                }
            }
        });
    }

    public void stop() {
        mListener = null;
    }

    // MARK: - Function
    private void doStartScan(AppCompatActivity activity) {
        DriverCardOCR.Companion.getShared().doScanCard(activity, new Function3<DriverCardOCR.RESULT, Bundle, DriverCardOCR.SCAN_TYPE, Unit>() {
            @Override
            public Unit invoke(DriverCardOCR.RESULT result, Bundle bundle, DriverCardOCR.SCAN_TYPE scanType) {
                if (result == DriverCardOCR.RESULT.SUCCESS) {
                    if (mListener != null) {
                        OcrResultInfo info = new OcrResultInfo();
                        if (scanType == DriverCardOCR.SCAN_TYPE.DriverCard) {
                            info.mCardType = CARDTYPE_MENKYOSYO;
                            if (bundle.containsKey("姓名")) {
                                info.mName = bundle.getString("姓名", "");
                            }
                            if (bundle.containsKey("住所")) {
                                info.mAddress = bundle.getString("住所", "");
                            }
                            if (bundle.containsKey("生年月日")) {
                                info.mBirthdate = bundle.getString("生年月日", "").replace("生", "");
                            }
                        } else if (scanType == DriverCardOCR.SCAN_TYPE.MyNumberCard) {
                            info.mCardType = CARDTYPE_MYNUMBER;
                            if (bundle.containsKey("氏名")) {
                                info.mName = bundle.getString("氏名", "");
                            }
                            String address = "";
                            if (bundle.containsKey("住所①")) {
                                address += bundle.getString("住所①", "");
                            }
                            if (bundle.containsKey("住所②")) {
                                address += bundle.getString("住所②", "");
                            }
                            info.mAddress = address;
                            if (bundle.containsKey("生年月日")) {
                                info.mBirthdate = bundle.getString("生年月日", "").replace("生", "");
                            }
                            if (bundle.containsKey("性別")) {
                                info.mGender = bundle.getString("性別", "");
                            }
                        } else if (scanType == DriverCardOCR.SCAN_TYPE.ZairyuCard) {
                            info.mCardType = CARDTYPE_ZAIRYU;
                            if (bundle.containsKey("氏名")) {
                                info.mName = bundle.getString("氏名", "");
                            }
                            if (bundle.containsKey("住所")) {
                                info.mAddress = bundle.getString("住所", "");
                            }
                            if (bundle.containsKey("生年月日")) {
                                info.mBirthdate = bundle.getString("生年月日", "").replace("生", "");
                            }
                            if (bundle.containsKey("性別")) {
                                info.mGender = bundle.getString("性別", "");
                            }
                        } else {
                            mListener.onResult(OcrPluginManager.this, CODE_FAILED, null);
                            return null;
                        }
                        mListener.onResult(OcrPluginManager.this, CODE_SUCCESS, info);

                    }
                }
                return null;
            }
        });
    }

    // MARK: - InnerClass
    public class OcrResultInfo implements Serializable {

        // MARK: - Member
        public int mCardType;
        public String mName;
        public String mAddress;
        public String mBirthdate;
        public String mGender;

    }

}
